### PR TITLE
Added support for Dameng user and role

### LINKED ISSUE #22396: Support for Dameng user and role
Support for Dameng user info and role info.
https://github.com/dbeaver/dbeaver/pull/22391
图片
图片

### PR DESCRIPTION
**Pull request by Stkai**
#22396
Support for view Dameng User info
Support for view Dameng Role info

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengPrivUser.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/*
+ * DBeaver - Universal Database Manager
+ * Copyright (C) 2010-2023 DBeaver Corp and others
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengPrivUser.java]: It is already 2024 :)

File: plugins/org.jkiss.dbeaver.ext.dameng/OSGI-INF/l10n/bundle.properties [hunk 2]
```diff
@@ -77,4 +86,20 @@ meta.org.jkiss.dbeaver.ext.dameng.model.DamengDataFile.maxSize.name=Max Size
...
 meta.org.jkiss.dbeaver.ext.dameng.model.DamengDataFile.nextSize.name=Next Size
 meta.org.jkiss.dbeaver.ext.dameng.model.DamengDataFile.mirrorPath.name=Mirror Path
-meta.org.jkiss.dbeaver.ext.dameng.model.DamengDataFile.realFreeSize.name=Real Free Size
\ No newline at end of file
+meta.org.jkiss.dbeaver.ext.dameng.model.DamengDataFile.realFreeSize.name=Real Free Size
+meta.org.jkiss.dbeaver.ext.dameng.model.DamengUser.objectId.name=User Id
+meta.org.jkiss.dbeaver.ext.dameng.model.DamengUser.name.name=Name
+meta.org.jkiss.dbeaver.ext.dameng.model.DamengUser.type.name=Type
+meta.org.jkiss.dbeaver.ext.dameng.model.DamengUser.tablespace.name=Tablespace
+meta.org.jkiss.dbeaver.ext.dameng.model.DamengUser.createTime.name=CreateTime
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.dameng/OSGI-INF/l10n/bundle.properties]:
```suggestion
meta.org.jkiss.dbeaver.ext.dameng.model.DamengUser.createTime.name=Create Time
```

File: plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengDataSource.java [hunk 3]
```diff
@@ -97,15 +102,88 @@ public static DBPDataKind getDataKind(@NotNull String typeName, int valueType) {
...
+                    "USER_OBJ.*,\n" +
+                    "SYSUSERS.*,\n" +
+                    "TS.ID AS TABLESPACE_ID,\n" +
+                    "TS.NAME AS TABLESPACE\n" +
+                    "FROM\n" +
+                    "SYSOBJECTS USER_OBJ\n" +
+                    "LEFT JOIN SYSUSERS ON\n" +
+                    "SYSUSERS.ID = USER_OBJ.ID\n" +
+                    "LEFT JOIN SYS.V$TABLESPACE TS ON\n" +
+                    "USER_OBJ.INFO3 & 0x000000000000FFFF = TS.ID\n" +
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengDataSource.java:177]: Could you please add a comment about this condition? It looks unusual.

**Comment by LonwoLonwo** [comment]
Hello @Stkai 

Thanks for your contribution.

Please create a ticket before the PR next time.
And now - just create a [ticket](https://github.com/dbeaver/dbeaver/issues) and add new changes with a ticket number in the commit.

File: plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengDataSource.java (hunk 3 above)
Comment (Stkai) [reply plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengDataSource.java:177]: When SYSOBJECTS is a user, bit(0-1) in info3 is the default tablespace id.

File: plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengPrivUser.java (hunk 1 above)
Comment (Stkai) [reply plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengPrivUser.java]: 2024, a new Year,the copyright has been updated

File: plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengDataSource.java (hunk 3 above)
Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengDataSource.java:177]: Ok. Could you please add it to the code? For the future developers :)

File: plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengUser.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    private Timestamp createTime;
+    private Type type;
+
+    private LockedStatus lockedStatus;
+
+    private long tablespaceId;
+
+    public DamengUser(DamengDataSource dataSource, JDBCResultSet dbResult) {
+        this.dataSource = dataSource;
+        this.id = JDBCUtils.safeGetLong(dbResult, "ID");
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengUser.java]: I think you can create some constants for the ID and NAME values. In the DamengConstants class.

**Comment by LonwoLonwo** [comment]
Thanks for the ticket creation.

You can add its ID to the following commits like this:
`#22396 Add constants`

All changes will be squashed during the merge so that we will see the ticket number in the annotation sections for your code.

File: plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengDataSource.java (hunk 3 above)
Comment (Stkai) [reply plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengDataSource.java:177]: Added :)

Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.ext.dameng/src/org/jkiss/dbeaver/ext/dameng/model/DamengDataSource.java:177]: Sorry, missed this one.

**Review by LonwoLonwo — APPROVED** [review]

**Comment by uslss** [comment]
verified

**Pull request merged**